main/src/main/java/com/airbnb/reair/incremental/configuration/SaslClusterUtils.java

```diff
@@ -0,0 +1,142 @@
+   */
+  public static void tokensInit(Credentials credentials, Configuration jobConfig)
+          throws TokenInitException {
+    try {
+      TokenCache.obtainTokensForNamenodes(
+          credentials,
+          new Path[] {
+              new Path(jobConfig.get(ConfigurationKeys.SRC_HDFS_ROOT)),
+              new Path(jobConfig.get(ConfigurationKeys.DEST_HDFS_ROOT)) },
+          jobConfig);
+
+      URI srcUri = new URI(jobConfig.get(ConfigurationKeys.SRC_CLUSTER_METASTORE_URL));
+      SaslClusterUtils.initMetastoreDelegationToken(
+          srcUri,
+          SaslClusterUtils.REAIR_KEY_TOKEN_SIGNATURE_SRC,
+          jobConfig.get(ConfigurationKeys.SRC_METASTORE_PRINCIPAL),
+          credentials
+      );
+
+      URI destUri = new URI(jobConfig.get(ConfigurationKeys.DEST_CLUSTER_METASTORE_URL));
+      SaslClusterUtils.initMetastoreDelegationToken(
+          destUri,
+          SaslClusterUtils.REAIR_KEY_TOKEN_SIGNATURE_DEST,
+          jobConfig.get(ConfigurationKeys.DEST_METASTORE_PRINCIPAL),
+          credentials
+      );
+    } catch (Exception ex) {
+      throw new TokenInitException(ex);
+    }
+  }
+
+  /**
+   * Create client for secured metastore service.
+   *
+   * @param uri An Uri to the metastore service
+   * @param signature Token id
+   * @param principal User principal
+   * @param conf Job configuration
+   * @return Metastore client
+   * @throws Exception If connecting to the metastore service fails
+   */
+  public static IMetaStoreClient getMetastoreClient(
+          URI uri, String signature, String principal, Configuration conf)
+          throws Exception {
+    HiveConf hiveConf = getHiveConf(conf, uri, principal);
+    String delegationToken = System.getenv("HADOOP_TOKEN_FILE_LOCATION");
+    if (delegationToken == null) {
+      FileSystem fs = FileSystem.get(conf);
+      Credentials creds = new Credentials();
+      fs.addDelegationTokens(UserGroupInformation.getCurrentUser().getUserName(), creds);
+      delegationToken = ".reair-" + System.currentTimeMillis() + ".dt";
+      Path filename = new Path(delegationToken);
+      creds.writeTokenStorageFile(filename, conf);
+      fs.deleteOnExit(filename);
+    }
+    hiveConf.set("mapreduce.job.credentials.binary", delegationToken);
+    hiveConf.set("hive.metastore.token.signature", signature);
+
+    return createMetastoreClient(uri, hiveConf);
+  }
+
+  private static void initMetastoreDelegationToken(
+          URI metastoreUri, String signature, String principal, Credentials credentials)
+          throws Exception {
+    Token<DelegationTokenIdentifier> token =
+            createMetastoreDelegationToken(metastoreUri, signature, principal);
+    credentials.addToken(token.getService(), token);
+    UserGroupInformation.getCurrentUser().addToken(token);
+  }
+
+  private static Token<DelegationTokenIdentifier> createMetastoreDelegationToken(
+          URI uri, String signature, String principal) throws Exception {
```

Comment:
It's good to avoid throwing the `Exceptions` since it's very general. Let's throw more specific exceptions?